### ISSUE TITLE
Major CPU performance decrease with ImageSharp.Web 3

### ISSUE THREAD

**Issue by marklagendijk** [SixLabors/ImageSharp.Web#344]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have verified that I am running the latest version of ImageSharp.Web
- [X] I have verified if the problem exist in both `DEBUG` and `RELEASE` mode
- [X] I have searched [open](https://github.com/SixLabors/ImageSharp.Web/issues) and [closed](https://github.com/SixLabors/ImageSharp.Web/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### ImageSharp.Web version

3.0.1

### Other Six Labors packages and versions

None

### Environment (Operating system, version and so on)

Azure App Service Windows

### .NET version

7

### Description

After upgrading to ImageSharp.Web 3 we noticed a major increase of CPU usage.
We then proceeded with steps to isolate the issue so we would be 100% sure that the version increase was indeed the cause:
1. We isolated the usage of ImageSharp.Web to its own application + App Service.
2. After a number of days we stepped down the version of ImageSharp.Web to version 2.
3. After around half a day I took a screenshot of the CPU performance (see below). I could wait longer, but the difference is so clear that I don't think that will help.

#### Technical background
* We run a multi-tenant web application that servers around 15 news websites.
* ImageSharp.Web is used for image downsizing.
* Azure FrontDoor is used as a pass-through CDN.
* Azure FrontDoor is configured to cache all the image responses for 1 year.
* Because Azure FrontDoor caches on the POP (point-of-presence) level, there are still numerous requests per image.
* Azure App Service (Windows, .NET 7) is used for serving the web applications. At this point we separated the usage of ImageSharp.Web to its own App Service, although we still run it on the same VM (Azure App Service Plan).
* We currently don't use caching on the ImageSharp.Web level, we configure a `NullImageCache`.
* Before we switched to the `NullImageCache` we used the `PhysicalFileSystemCache`. It ended up breaking our App Service because we eventually ran out of disk space, because the cache does not remove old images (note: I now know that this is a non-trivial issue, I would propose making this more clear in the documentation).
* The images are loaded with a custom `RemoteImageProvider` which uses `HttpClient` to load images from an url.
* We use the `PresetOnlyQueryCollectionRequestParser`.

#### Background
A few weeks back I found out that version 3 of ImageSharp.Web was released. I also found out that the license had been changed.
After arranging the purchase of the license I upgraded.

Based on the release notes I expected to see a performance improvement, but unfortunately it quickly became clear that we instead had a performance decrease.

I explicitly waited with reporting the issue until I was 100% sure that it is caused by the ImageSharp.Web upgrade.
I checked the changelogs of ImageSharp.Web and ImageSharp, but I could not find any change that which might cause our issue.

I am willing to help with identifying the cause of the issue, but I would like some pointers with regards to how to proceed.

### Steps to Reproduce

At this point the "Steps to Reproduce" are not clear beyond "Upgrade from ImageSharp.Web 2.0.2 to 3.0.1".
However, I can share all the relevant code that we use.

This is the library that contains all our `ImageSharp.Web` code.
[PMP.ImageProcessing.zip](https://github.com/SixLabors/ImageSharp.Web/files/13248304/PMP.ImageProcessing.zip)

The `PMP.ImageProcessing` library is used in a minimal web application.
[PMP.ImageProcessor.zip](https://github.com/SixLabors/ImageSharp.Web/files/13248321/PMP.ImageProcessor.zip)

### Images

![Screenshot-Metrics - Microsoft Azure — Mozilla Firefox-2023-11-03 09_17_21](https://github.com/SixLabors/ImageSharp.Web/assets/1188039/aa2e1cbc-4282-45b2-8220-fd454fa88930)

The red arrow indicates the point where the version 2 of ImageSharp.Web was deployed.
The graph shows that with ImageSharp.Web 3 both average and maximum CPU usage are at least 2 times higher.

**Comment by JimBobSquarePants** [comment]
Thanks for the detail. Have you been able to profile this? 

I’m wondering if this is related? It might be worth trying one of the nightly builds. Hard to determine the issue without profiling. 

https://github.com/SixLabors/ImageSharp.Web/pull/327

**Comment by marklagendijk** [comment]
I have not been able to profile the issue.

Is the nightly build available via NuGet, or do I need to build it myself?

We don't use the HMAC-protection. Is it enabled by default, if so, is there a way to disable it?

**Comment by JimBobSquarePants** [comment]
I've been doing some profiling comparing the two versions and I cannot see anything significant. Are you experiencing a drop in throughput? 

<img alt="image" src="https://github.com/SixLabors/ImageSharp.Web/assets/385879/42300739-f399-4b6e-a727-4ddf1550061c">

**Comment by marklagendijk** [comment]
I did try the nightly build.
It did not seem to make any difference.

Based on your comment I compared the `Response Time` metric for the 2.x and 3.x. 
It is significantly higher (worse performance) for 3.x vs 2.x:
- 50th percentile: 47ms vs 36ms
- 95th percentile: 280ms vs 220ms
- 99th percentile: 840ms vs 410ms
- Average: 100ms vs 70ms

Now I do have a license for all Jetbrains .NET tools, including the profiler build into Rider and the separate dotTrace Profiler application, however I don't have much experience with it.
I could quite easily get to the point where I can run a fixed set of image processor requests. I think this might be useful because it could be that the issue only applies to certain types of requests, so doing a set instead of a single one could help pin down the issue.
Is it easy to use the profiling tool to compare such a set of requests?

**Comment by JimBobSquarePants** [comment]
The Jetbrains Tools don't really provide the detailed information you need running under stress so instead I've benchmarked the solution using [Crank](https://github.com/dotnet/crank)

It's fairly straightforward to get started. 

1. First install the global tool as per the [instructions.](https://github.com/dotnet/crank/blob/main/docs/getting_started.md)
2. Run `crank-agent`
3. Navigate to each folder in the zip file in turn and run the following, letting each run complete before starting the other. 
`crank --config local.benchmarks.yml --scenario imagesharp --profile local`

[crank.zip](https://github.com/SixLabors/ImageSharp.Web/files/13335072/crank.zip)

Now... That all said I'm not seeing the issues you are,

Application build and startup churns through the CPU a bit more for some reason but the load tests are equivalent. If anything v3 has a slightly higher max throughput. 

<img alt="image" src="https://github.com/SixLabors/ImageSharp.Web/assets/385879/2143c483-5956-46ff-a969-78dd49539051">

**Comment by adamsitnik** [comment]
My 2 cents: 

1. How does the CPU spike relates to the number of requests being handled by the web service? Is it possible that a single instance of a web service is now handling more requests?
2. What is the CPU specification? Is @JimBobSquarePants running the benchmarks on the same architecture (x64 vs arm64), with the same vector size and available instruction set?

**Comment by marklagendijk** [comment]
I just (yesterday) upgraded to the newly released 3.1 version, and the problematic CPU usage seems to be fixed.
I'll monitor the situation for a few days to be absolutely sure.

**Comment by marklagendijk** [comment]
I can happily report that our application has been running 3.1 for several days now and has had no CPU issues.
With this I conclude that the original issue (whatever it may have been) has been fixed since 3.0.

**Comment by marklagendijk** [comment]
@JimBobSquarePants I seem to be running into the issue described in https://github.com/SixLabors/ImageSharp/issues/2450.
I see that you fixed that issue and deployed it in ImageSharp 3.1 (which is also part of ImageSharp.Web 3.1).

Do I need to configure the GIF Encoder with specific settings to use this, or may it be that I'm running into a different GIF issue?

**Comment by marklagendijk** [comment]
Sure. We have the issue with GIF advertisement banners.
This is one example:

![Havecon2022](https://github.com/SixLabors/ImageSharp.Web/assets/1188039/3ccc9a7d-1262-4dae-a450-06d1360d4966)
![Havecon2022-broken](https://github.com/SixLabors/ImageSharp.Web/assets/1188039/adfbf8d8-8ec9-43ee-8f74-e9e420729144)

It seems to happen with a significant percentage of our banners.

**Comment by marklagendijk** [comment]
Awesome! That is a size reduction of 26%!
Given that most of our hosting costs are bandwidth (Azure FrontDoor bandwidth is not cheap) this is very significant for us.

When it's out I will verify it locally against all our current banners.

**Comment by JimBobSquarePants** [comment]
You can grab ImageSharp v3.1.1 now from NuGet. It's binary compatible with ImageSharp.Web v3.1.0

If your company hasn't purchased a license already and meet the criterea, please do so. It helps keep the lights on here.

**Comment by marklagendijk** [comment]
Wow, that was quick! 🚀 
Btw, yes, we did buy a license.

I can confirm that this fixes the distorted GIF issue.

When comparing file sizes I did notice a case where the processed image was _bigger_ than the original:
Processed:
![DelTomSky23-processed](https://github.com/SixLabors/ImageSharp.Web/assets/1188039/2bc77455-4ca1-4442-834a-69e33c2d9bf6)

Original:
![DelTomSky23-original](https://github.com/SixLabors/ImageSharp.Web/assets/1188039/f759fef2-45a0-4984-aa0b-7e885e2105d7)

I will create a script to automatically compare the file sizes of all our images to see the distribution of size difference over the entire dataset.
Is it reasonable to expect that the output size of ImageSharp.Web will always be smaller or the same as the input (assuming no size-increasing operations, of course)? Or is this not feasible due to complicated technical details?

**Comment by JimBobSquarePants** [comment]
There’s no guarantees. When we load a gif we have to coalesce the frame deltas losing all previous optimisations. We do a ton of work to reintroduce optimisations but sometimes we can’t match original meticulous individual encodings. There are changes we could theoretically make to the deduping process to enable better compression but they’d make encoding times too slow.

**Comment by marklagendijk** [comment]
I get that. 

For the web use case I think it would be nice to have the option to use more aggressive compression. 
For our case it could be totally acceptable to have processing times be up to 10 times higher if that would help with reducing image size, the result is cached after all.

It's not even only about saving bandwidth costs, but also about faster load times for the end user, and saving them bandwidth.

**Comment by JimBobSquarePants** [comment]
It’s very intense for gif since there’s no line filtering. You’d have to have multiple deduplication strategies that ignore diffs when there are not multiple in a row. Then you’d have to compress each whole image frame multiple times and compare. 

Ideally our current strategy saves overall so individual differences become less relevant.
